### PR TITLE
Added VGG19 Model for MNIST Dataset.

### PR DESCRIPTION
**Pull request by MuLx10**
VGG19 Model for MNIST.
It uses Kaggle MNIST dataset

### CODE REVIEW

**Comment by zoq** [comment]
Nice, do you have some first results for the model?

**Comment by MuLx10** [comment]
I don't have any results as of now. I ran for 1 cycle to check if the code was working and model was being saved correctly.

**Comment by zoq** [comment]
I see, can you run the model for a couple of iterations or should I start the training process?

**Comment by MuLx10** [comment]
> I see, can you run the model for a couple of iterations or should I start the training process?

I have started the training.

**Comment by sumedhghaisas** [comment]
Excited to see the results. :) How long did 1 iteration took? Just for a ballpark measure

File: VGG19/vgg19_mnist/src/VGG19.hpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+#include <mlpack/core.hpp>
+#include <mlpack/prereqs.hpp>
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/ann/ffn.hpp>
+
+using namespace mlpack;
+using namespace mlpack::ann;
+
+class VGG19
```
Comment (zoq) [inline VGG19/vgg19_mnist/src/VGG19.hpp]: We should add some more information about the model here, a reference to the paper would also be helpful.

File: VGG19/vgg19_mnist/src/VGG19.hpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+{
+public:
+
+  VGG19(const size_t inputWidth,
```
Comment (zoq) [inline VGG19/vgg19_mnist/src/VGG19.hpp]: A description for each parameter would be nice to have.

File: VGG19/vgg19_mnist/src/VGG19.hpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        const size_t inputHeight,
+        const size_t inputChannel,
+        const size_t numClasses,
+        const bool includeTop = true,
+        const std::string pooling = "max",
+        const std::string weights = "imagenet"
```
Comment (zoq) [inline VGG19/vgg19_mnist/src/VGG19.hpp]: Not sure I get the meaning behind the parameter looks like it's unused.

File: VGG19/vgg19_mnist/src/VGG19.hpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        );
```
Comment (zoq) [inline VGG19/vgg19_mnist/src/VGG19.hpp]: This should end right behind the laster parameter.

File: VGG19/vgg19_mnist/src/VGG19_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+#include "VGG19.hpp"
+
+VGG19::VGG19(const size_t inputWidth,
+             const size_t inputHeight,
+             const size_t inputChannel,
+             const size_t numClasses,
+             const bool includeTop,
+             const std::string pooling,
+             const std::string weights
+            ):
```
Comment (zoq) [inline VGG19/vgg19_mnist/src/VGG19_impl.hpp]: See comment above.

File: VGG19/vgg19_mnist/src/VGG19_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        2,  // Width of field.
+        2,  // Height of field.
+        2,  // Stride along width.
+        2,  // Stride along height.
+        true
+      );
+
+    std::cout << "Width" << " " << "Height" << std::endl;
+
+    // same padding
```
Comment (zoq) [inline VGG19/vgg19_mnist/src/VGG19_impl.hpp]: Please use camel casing for all names.

File: VGG19/vgg19_mnist/src/VGG19.hpp (hunk 3 above)
Comment (MuLx10) [reply VGG19/vgg19_mnist/src/VGG19.hpp]: Pooling is for choosing between global max/avg pooling.
weights parameter would make the VGG load imagenet trained weights, else it would make VGG start from scratch. Since it is mnist based hence not used
.

**Comment by zoq** [comment]
Any updates on the results?

File: VGG19/vgg19_mnist/src/VGG19_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+    }
+    return VGGNet;
+}
+
+size_t VGG19::GetOutputShape()
+{
+    return outputShape;
+}
+
+Sequential<>* VGG19::LoadModel(std::string filePath)
```
Comment (zoq) [inline VGG19/vgg19_mnist/src/VGG19_impl.hpp]: What about using ``const std::string& ``here and for the save method as well; avoids another copy.

**Comment by MuLx10** [comment]
> Any updates on the results?

Four cycles completed.
Epoch 0
Loss after cycle 0 -> 311.907
Epoch 1
Loss after cycle 1 -> 311.83
Epoch 2
Loss after cycle 2 -> 311.731
Epoch 3
Loss after cycle 3 -> 311.658
Epoch 4
Loss after cycle 4 -> 311.349

**Comment by zoq** [comment]
Hm, maybe it makes sense to start on a subset to figure out some reasonable optimizer settings, looking at the first iterations I would increase the step size.

**Comment by sumedhghaisas** [comment]
What step size did you use for this training? I agree with Marcus here to increase the step size.

**Comment by MuLx10** [comment]
Initial step size was 1.2e-3. I changed it to 5e-3 and restarted the training. Should I increase it more?

**Comment by sumedhghaisas** [comment]
hmm... thats interesting. Usually the VGG models (smaller ones) I train require 1e-3 step size. I checked the Adam configuration you used and that should do it. For bigger models higher step size require higher batch size, or batch normalization. How does the rest of the training curve look like? Does the loss go down later?

**Comment by sumedhghaisas** [comment]
Another thing you could try is using high learning rate in the start but decrease it later.

**Comment by zoq** [comment]
I would also test another weight initialization policy.

**Comment by zoq** [comment]
@MuLx10 Any update on this one?

**Comment by MuLx10** [comment]
New paramters (lr = 5e-3) caused divergence.
Epoch 0
Loss after cycle 0 -> 310.261
Epoch 1
Loss after cycle 1 -> 223.393
Epoch 2
Loss after cycle 2 -> 2.83091e+10
Epoch 3
Loss after cycle 3 -> 6.03322e+45
Epoch 4
Loss after cycle 4 -> 1.6653e+49
Epoch 5
Loss after cycle 5 -> 3.37138e+23
Epoch 6
Loss after cycle 6 -> 8.95541e+53

File: VGG19/vgg19_mnist/src/VGG19_impl.hpp [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+        // Add the final dense layer.
+        VGGNet->Add<Linear<> >(inputWidth * inputHeight * 512, 4096);
+        VGGNet->Add<Linear<> >(4096, 4096);
+        VGGNet->Add<Linear<> >(4096, numClasses);
+        outputShape = numClasses;
+    }
+    else
+    {   
+        outputShape = inputWidth * inputHeight * 512;
+        if (pooling == "max")
```
[the remark this answers is not in the document]
Comment (MuLx10) [reply VGG19/vgg19_mnist/src/VGG19_impl.hpp]: Done as you told.

**Comment by zoq** [comment]
Okay, do you plan to test the ideas mentioned above next?

**Comment by MuLx10** [comment]
I have restarted the training with different initialization and step size.

**Comment by zoq** [comment]
Btw. if you don't see reasonable results in the first 10 iterations, you can stop the training process; in this case there might be something wrong with the model, in which case let's create a super simple subset (like 50 samples) and test on that one.

File: VGG19/vgg19_mnist/src/model.cpp [hunk 9]
```diff
@@ -82,35 +83,34 @@ int main()
...
+  X /= 255.0;
+  // Create labels for the datasets.
+  data::OneHotEncoding(arma::conv_to<arma::irowvec>::from(dataset.row(0)), Y);
+  Y = Y.t();
 
-  // The train and valid datasets contain both - the features as well as the
-  // class labels. Split these into separate mats.
-  const mat trainX = train.submat(1, 0, train.n_rows - 1, train.n_cols - 1);
-  const mat validX = valid.submat(1, 0, valid.n_rows - 1, valid.n_cols - 1);
+  // Split the dataset into training and validation sets.
```
Comment (zoq) [inline VGG19/vgg19_mnist/src/model.cpp]: You could use: https://github.com/mlpack/mlpack/blob/master/src/mlpack/core/data/split_data.hpp for the splitting as well.

Comment (MuLx10) [reply VGG19/vgg19_mnist/src/model.cpp]: Sure.

**Comment by zoq** [comment]
I think once we see good results (is there any update on that point?), this is ready to be merged. A simple tutorial about the model would be nice to have.

**Comment by MuLx10** [comment]
Sure I'll add it.

File: VGG19/vgg19_mnist/src/model.cpp [hunk 10]
```diff
@@ -10,7 +10,7 @@
  * http://www.opensource.org/licenses/BSD-3-Clause for more information.
  */
 
-#include "VGG19.hpp"
+#include "../../VGG19.hpp"
```
Comment (zoq) [inline VGG19/vgg19_mnist/src/model.cpp]: Nice catch!

**Comment by MuLx10** [comment]
> Any update on the tutorial part?

I'll add tutorial for both (mnist & imagenet) the example together.

**Comment by zoq** [comment]
Any update on that? Maybe CIFAR 10 is another solution that is way faster to train.

**Comment by zoq** [comment]
Thanks for adding the tutorial, hopefully we see some good results.
